Source/Plugin.BLE/Android/Device.cs

```diff
@@ -262,58 +263,68 @@ public static List<AdvertisementRecord> ParseScanRecord(byte[] scanRecord)
-                    case AdvertisementRecordType.SsUuids16Bit:
-                    case AdvertisementRecordType.SsUuids32Bit:
-                    case AdvertisementRecordType.UuidCom32Bit:
-                    case AdvertisementRecordType.UuidsComplete128Bit:
-                    case AdvertisementRecordType.UuidsComplete16Bit:
-                    case AdvertisementRecordType.UuidsIncomple16Bit:
-                    case AdvertisementRecordType.UuidsIncomplete128Bit:
-                        Array.Reverse(data);
+                    if (!Enum.IsDefined(typeof(AdvertisementRecordType), type))
+                    {
+                        Trace.Message("Advertisment record type not defined: {0}", type);
                         break;
-                }
-                var record = new AdvertisementRecord((AdvertisementRecordType)type, data);
+                    }
 
-                Trace.Message(record.ToString());
+                    //data length is length -1 because type takes the first byte
+                    byte[] data = new byte[length - 1];
+                    Array.Copy(scanRecord, index + 1, data, 0, length - 1);
 
-                records.Add(record);
+                    // don't forget that data is little endian so reverse
+                    // Supplement to Bluetooth Core Specification 1
+                    // NOTE: all relevant devices are already little endian, so this is not necessary for any type except UUIDs
+                    //var record = new AdvertisementRecord((AdvertisementRecordType)type, data.Reverse().ToArray());
 
-                //Advance
-                index += length;
+                    switch ((AdvertisementRecordType)type)
+                    {
+                        case AdvertisementRecordType.ServiceDataUuid32Bit:
+                        case AdvertisementRecordType.SsUuids128Bit:
+                        case AdvertisementRecordType.SsUuids16Bit:
+                        case AdvertisementRecordType.SsUuids32Bit:
+                        case AdvertisementRecordType.UuidCom32Bit:
+                        case AdvertisementRecordType.UuidsComplete128Bit:
+                        case AdvertisementRecordType.UuidsComplete16Bit:
+                        case AdvertisementRecordType.UuidsIncomple16Bit:
+                        case AdvertisementRecordType.UuidsIncomplete128Bit:
+                            Array.Reverse(data);
+                            break;
+                    }
+                    var record = new AdvertisementRecord((AdvertisementRecordType)type, data);
+
+                    Trace.Message(record.ToString());
+
+                    records.Add(record);
+
+                    //Advance
+                    index += length;
+                }
+                return records;
+            }
+            catch(Exception)
+            {
+                Trace.Message($"Filed to parse advertisement data: {scanRecord.ToHexString()}");
```

Comment:
Typo here: "Filed" -> "Failed"

Reply:
In addition: Maybe your trace output should also contain the device address, so that it is easier to see which device is causing the issue?

Reply:
Yes you are right

Reply:
How would you like to pass the device address to the function?
For example like this:
~~~C#
public static List<AdvertisementRecord> ParseScanRecord(byte[] scanRecord, BluetoothDevice nativeDevice)
{
   // ...
   Trace.Message("Failed to parse advertisementData: {0} device address: {1}", scanRecord.ToHexString(), nativeDevice.Address);
}
~~~

Reply:
> How would you like to pass the device address to the function?

This is not necessary. The `Device` class has a member `NativeDevice`, so you can directly access `NativeDevice.Address` in the method `ParseScanRecord`.